snapcraft/internal/pluginhandler/stage_package_grammar/_on.py

```diff
@@ -0,0 +1,128 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from . import (
+    process_grammar,
+    StagePackageSyntaxError
+)
+
+
+class OnStatement:
+    """Process an 'on' statement in the stage packages grammar.
+
+    Example:
+    >>> import tempfile
+    >>> from snapcraft import repo, ProjectOptions
+    >>> with tempfile.TemporaryDirectory() as cache_dir:
+    ...     ubuntu = repo.Ubuntu(cache_dir)
+    ...     options = ProjectOptions(target_deb_arch='i386')
+    ...     clause = OnStatement('on amd64', ['foo'], options, ubuntu)
+    ...     clause.add_else(['bar'])
+    ...     clause.process()
+    {'bar'}
+    """
+
+    def __init__(self, on, body, project_options, ubuntu):
```

Comment:
True, keyword-only removes any ordering from the function call, which is nice, but of course has the downside of requiring more typing.  What you're suggesting, though, has broader implications. Should we start using keyword-only in every function, then? Not even golang is that verbose :wink: . Given that you're bringing this from `golint` and `struct`s, do you really feel this way only about `__init__()`?